### PR TITLE
replication: pass checksum headers to replica

### PR DESCRIPTION
**Pull request by poornas**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Motivation and Context
replication needs to preserve crc/sha256 checksum headers on replica when set.

## How to test this PR?
upload an object with checksum headers, then do a head with checksum-mode enabled
```
aws s3api head-object --bucket bucket --key dzzzmlka4vunwvyp2blfvmwnd5wjyl --profile minio --endpoint-url https://localhost:9002 --no-verify-ssl --checksum-mode ENABLED
/opt/homebrew/lib/python3.9/site-packages/urllib3/connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "AcceptRanges": "bytes",
    "LastModified": "Wed, 29 May 2024 21:17:29 GMT",
    "ContentLength": 10240,
    "ChecksumSHA1": "bDkbZ38UR2O9+gsg1BwS0Oquf1s=",
    "ETag": "\"a0f9fa4ae8c9c3fc0444af597dd73d7c\"",
    "VersionId": "72b044c6-717a-4dbd-8f82-c3f82f661f65",
    "ContentType": "application/octet-stream",
    "Metadata": {},
    "ReplicationStatus": "REPLICA"
}
➜  scripts aws s3api head-object --bucket bucket --key dzzzmlka4vunwvyp2blfvmwnd5wjyl --profile minio --endpoint-url https://localhost:9004 --no-verify-ssl --checksum-mode ENABLED
/opt/homebrew/lib/python3.9/site-packages/urllib3/connectionpool.py:1043: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
  warnings.warn(
{
    "AcceptRanges": "bytes",
    "LastModified": "Wed, 29 May 2024 21:17:29 GMT",
    "ContentLength": 10240,
    "ChecksumSHA1": "bDkbZ38UR2O9+gsg1BwS0Oquf1s=",
    "ETag": "\"a0f9fa4ae8c9c3fc0444af597dd73d7c\"",
    "VersionId": "72b044c6-717a-4dbd-8f82-c3f82f661f65",
    "ContentType": "application/octet-stream",
    "Metadata": {},
    "ReplicationStatus": "COMPLETED"
}
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/bucket-replication.go [hunk 1]
```diff
@@ -769,6 +769,20 @@ func putReplicationOpts(ctx context.Context, sc string, objInfo ObjectInfo) (put
...
+							cksum := hash.NewChecksumString(k, v)
+							if cksum == nil {
+								continue
+							}
+							if cksum.Valid() {
+								meta[cksum.Type.Key()] = v
+							}
+						}
+					}
+				}
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]: this is not the only checksum @poornas we have support for sha1, sha256 and crc32c?

**Review by harshavardhana — CHANGES_REQUESTED** [review]
Please add a test case here.

File: cmd/bucket-replication.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-replication.go]: https://github.com/minio/minio/blob/master/internal/hash/checksum.go#L91 - all of it is covered

Comment (harshavardhana) [reply cmd/bucket-replication.go]: does this cover multipart checksums as well?

File: cmd/bucket-replication.go [hunk 2]
```diff
@@ -769,6 +769,20 @@ func putReplicationOpts(ctx context.Context, sc string, objInfo ObjectInfo) (put
 		// In case of SSE-C objects copy the allowed internal headers as well
```
Comment (harshavardhana) [inline cmd/bucket-replication.go:784]: will sse-C preserve these headers as is?

File: cmd/bucket-replication.go [hunk 3]
```diff
@@ -769,6 +769,20 @@ func putReplicationOpts(ctx context.Context, sc string, objInfo ObjectInfo) (put
...
 		if !crypto.SSEC.IsEncrypted(objInfo.UserDefined) || !slices.Contains(maps.Keys(validSSEReplicationHeaders), k) {
 			if stringsHasPrefixFold(k, ReservedMetadataPrefixLower) {
+				if strings.EqualFold(k, ReservedMetadataPrefixLower+"crc") {
+					cs := objInfo.decryptChecksums(0)
+					if cs != nil {
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]:
```suggestion
					if cs != nil {
```

superflous not needed just do `for k, v := range cs`

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/bucket-replication.go (hunk 1 above)
Comment (poornas) [reply cmd/bucket-replication.go]: you raise interesting questions - do not know how SSEC / multipark checksums need to be handled. will do some testing , converting this to draft PR for now.

**Review by harshavardhana — APPROVED** [review]

**Comment by shtripat** [comment]
Tested normal upload and multipart upload in HTTP mode and works as expected. But for SSEC enabled HTTPS but normal object upload itself doesn't show checksum in `head-object` as below. Am I missing something here? While `put-object` response does contain the checksum though.

```
$ mc alias set m1 https://localhost:9000/ minioadmin minioadmin --insecure
Added `m1` successfully.
$ mc alias set m2 https://localhost:9010/ minioadmin minioadmin --insecure
Added `m2` successfully.
$ mc mb m1/test-bucket --insecure
Bucket created successfully `m1/test-bucket`.
$ mc admin replicate add m1 m2 --insecure
Requested sites were configured for replication successfully.
$ mc ls m1 --insecure
[2024-06-06 10:51:16 IST]     0B test-bucket/
$ mc ls m2 --insecure
[2024-06-06 10:51:16 IST]     0B test-bucket/
$ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY=",
    "VersionId": "80663209-5eb1-4698-953f-daeff64f7023"
}
$ mc ls m1/test-bucket --insecure
[2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
$ mc ls m2/test-bucket --insecure
[2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --checksum-mode ENABLED --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-06T05:22:54+00:00",
    "ContentLength": 16,
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "ReplicationStatus": "COMPLETED"
}
$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9010/ --checksum-mode ENABLED --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-06T05:22:54+00:00",
    "ContentLength": 16,
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "ReplicationStatus": "REPLICA"
}
```

**Comment by shtripat** [comment]
> --profile minio

This doesn't seem a replication issue as simple load to a HTTPS enabled MinIO as well doesnt show the checksum in `head-object` response
```
$ mc mb m1/test-bucket --insecure
Bucket created successfully `m1/test-bucket`.

$ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY="
}

$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --checksum-mode ENABLED --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-06T06:23:24+00:00",
    "ContentLength": 16,
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "ContentType": "binary/octet-stream",
    "Metadata": {}
}
```

**Comment by shtripat** [comment]
> Tested normal upload and multipart upload in HTTP mode and works as expected. But for SSEC enabled HTTPS but normal object upload itself doesn't show checksum in `head-object` as below. Am I missing something here? While `put-object` response does contain the checksum though.
> 
> ```
> $ mc alias set m1 https://localhost:9000/ minioadmin minioadmin --insecure
> Added `m1` successfully.
> $ mc alias set m2 https://localhost:9010/ minioadmin minioadmin --insecure
> Added `m2` successfully.
> $ mc mb m1/test-bucket --insecure
> Bucket created successfully `m1/test-bucket`.
> $ mc admin replicate add m1 m2 --insecure
> Requested sites were configured for replication successfully.
> $ mc ls m1 --insecure
> [2024-06-06 10:51:16 IST]     0B test-bucket/
> $ mc ls m2 --insecure
> [2024-06-06 10:51:16 IST]     0B test-bucket/
> $ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
> urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> {
>     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
>     "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY=",
>     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023"
> }
> $ mc ls m1/test-bucket --insecure
> [2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
> $ mc ls m2/test-bucket --insecure
> [2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
> $ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --checksum-mode ENABLED --no-verify-ssl
> urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> {
>     "AcceptRanges": "bytes",
>     "LastModified": "2024-06-06T05:22:54+00:00",
>     "ContentLength": 16,
>     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
>     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
>     "ContentType": "binary/octet-stream",
>     "Metadata": {},
>     "ReplicationStatus": "COMPLETED"
> }
> $ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9010/ --checksum-mode ENABLED --no-verify-ssl
> urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> {
>     "AcceptRanges": "bytes",
>     "LastModified": "2024-06-06T05:22:54+00:00",
>     "ContentLength": 16,
>     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
>     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
>     "ContentType": "binary/octet-stream",
>     "Metadata": {},
>     "ReplicationStatus": "REPLICA"
> }
> ```

This doesn't seem a replication issue as simple load to a HTTPS enabled MinIO as well doesnt show the checksum in head-object response

```
$ mc mb m1/test-bucket --insecure
Bucket created successfully `m1/test-bucket`.

$ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY="
}

$ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --checksum-mode ENABLED --no-verify-ssl
urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
{
    "AcceptRanges": "bytes",
    "LastModified": "2024-06-06T06:23:24+00:00",
    "ContentLength": 16,
    "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
    "ContentType": "binary/octet-stream",
    "Metadata": {}
}
```

**Comment by shtripat** [comment]
> --no-verify-ssl



> > Tested normal upload and multipart upload in HTTP mode and works as expected. But for SSEC enabled HTTPS but normal object upload itself doesn't show checksum in `head-object` as below. Am I missing something here? While `put-object` response does contain the checksum though.
> > ```
> > $ mc alias set m1 https://localhost:9000/ minioadmin minioadmin --insecure
> > Added `m1` successfully.
> > $ mc alias set m2 https://localhost:9010/ minioadmin minioadmin --insecure
> > Added `m2` successfully.
> > $ mc mb m1/test-bucket --insecure
> > Bucket created successfully `m1/test-bucket`.
> > $ mc admin replicate add m1 m2 --insecure
> > Requested sites were configured for replication successfully.
> > $ mc ls m1 --insecure
> > [2024-06-06 10:51:16 IST]     0B test-bucket/
> > $ mc ls m2 --insecure
> > [2024-06-06 10:51:16 IST]     0B test-bucket/
> > $ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
> > urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> > {
> >     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
> >     "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY=",
> >     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023"
> > }
> > $ mc ls m1/test-bucket --insecure
> > [2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
> > $ mc ls m2/test-bucket --insecure
> > [2024-06-06 10:52:54 IST]    16B STANDARD hello.txt
> > $ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000/ --checksum-mode ENABLED --no-verify-ssl
> > urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> > {
> >     "AcceptRanges": "bytes",
> >     "LastModified": "2024-06-06T05:22:54+00:00",
> >     "ContentLength": 16,
> >     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
> >     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
> >     "ContentType": "binary/octet-stream",
> >     "Metadata": {},
> >     "ReplicationStatus": "COMPLETED"
> > }
> > $ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9010/ --checksum-mode ENABLED --no-verify-ssl
> > urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> > {
> >     "AcceptRanges": "bytes",
> >     "LastModified": "2024-06-06T05:22:54+00:00",
> >     "ContentLength": 16,
> >     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
> >     "VersionId": "80663209-5eb1-4698-953f-daeff64f7023",
> >     "ContentType": "binary/octet-stream",
> >     "Metadata": {},
> >     "ReplicationStatus": "REPLICA"
> > }
> > ```
> 
> This doesn't seem a replication issue as simple load to a HTTPS enabled MinIO as well doesnt show the checksum in head-object response
> 
> ```
> $ mc mb m1/test-bucket --insecure
> Bucket created successfully `m1/test-bucket`.
> 
> $ aws s3api put-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --body hello.txt --checksum-algorithm SHA256 --no-verify-ssl
> urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> {
>     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
>     "ChecksumSHA256": "m8Z5HigGuc6eY/VXKMUZkYCJdu8bd7kuFqF2vUQ81IY="
> }
> 
> $ aws s3api head-object --bucket test-bucket --key hello.txt --endpoint-url https://localhost:9000 --checksum-mode ENABLED --no-verify-ssl
> urllib3/connectionpool.py:1061: InsecureRequestWarning: Unverified HTTPS request is being made to host 'localhost'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#ssl-warnings
> {
>     "AcceptRanges": "bytes",
>     "LastModified": "2024-06-06T06:23:24+00:00",
>     "ContentLength": 16,
>     "ETag": "\"8a920fae3b35ae00d11f9beade4634e9\"",
>     "ContentType": "binary/octet-stream",
>     "Metadata": {}
> }
> ```

So this is found to be a aws cli thingy. We explicitly pass calculated `--checksum-sha256` as well with `--checksum-algorithm SHA256` it works fine. In case of https, aws cli is not auto calculating and sending `--checksum-sha256`

**Comment by harshavardhana** [comment]
That is to be expected @shtripat because with tls enabled they are not calculating payload checksums since TLS provides payload protection.

**Comment by harshavardhana** [comment]
Can you approve this PR?

**Comment by shtripat** [comment]
> Can you approve this PR?

Approving for now. Still need to test ssec objects though

**Review by shtripat — APPROVED** [review]

**Pull request merged**